pkg/certificate/certificate.go

```diff
@@ -0,0 +1,415 @@
+// Copyright Contributors to the Open Cluster Management project
+package certificate
+
+import (
+	"crypto/x509"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/spf13/pflag"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/klog"
+	netutils "k8s.io/utils/net"
+	"open-cluster-management.io/multicluster-controlplane/pkg/certificate/certchains"
+	"open-cluster-management.io/multicluster-controlplane/pkg/servers/options"
+	"open-cluster-management.io/multicluster-controlplane/pkg/util"
+)
+
+var DefaultMulticlusterControlplaneConfigDir = ".ocmconfig"
+
+func NewMulticlusterCertificateConfig() *MulticlusterCertificateConfig {
+	return &MulticlusterCertificateConfig{}
+}
+
+type MulticlusterCertificateConfig struct {
+	// these variables are used to sign certs
+	ApiHost   string
+	ApiHostIP string
+	URL       string
+	// command line flags
+	GenerateCertificate bool
+	DeployMode          bool
+	useEmbedEtcd        bool
+	// config directory
+	ConfigDir string
+}
+
+func (cfg *MulticlusterCertificateConfig) AddFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&cfg.GenerateCertificate, "generate-cert", false, "If set, the controlplane will auto-generate certificates.")
+	fs.BoolVar(&cfg.DeployMode, "deploy-mode", false, "If set, means we deploy multicluster controlplane to an ocp cluster")
+	fs.StringVar(&cfg.ConfigDir, "config-dir", DefaultMulticlusterControlplaneConfigDir, "Directory to store all configs.")
+}
+
+func (cfg *MulticlusterCertificateConfig) InitCertsForServerRunOptions(o *options.ServerRunOptions) *options.ServerRunOptions {
+	if cfg.DeployMode {
+		// if in deploy mode, apihost should be set
+		cfg.ApiHost = o.GenericServerRunOptions.ExternalHost
+	} else {
```

Comment:
In AMI image(starting an EC2 instance and run control-plane as a systemd service)case, control-plane wants to be deployed  in non deploy mode, but we want to specify the API host.
Does the current logic cover this case?

/cc @clyang82

Reply:
Yes, the API host/IP needs to put to the SAN of certificate.
In AMI case, the `deployMode` should set be `false`, right? because controlplane is running as a systemd service, and `kubeconfig` will be written into a path instead of a k8s secret. 
API host/IP in AMI case should be a public IP(public IP for EC2 instance) that can't be got from `util.GetLocalIP()`, we should manually specify the `API host/IP` in controlplane's configuration.

From this code logic, `non-deployMode` and `API host/IP` are exclusive, it will not cover AMI case.

Reply:
As an alternative, we don't need to change this.
I will update `util.GetLocalIP()` to add logic to get public IP of EC2 instance firstly.